FLT/AutomorphicForm/QuaternionAlgebra/HeckeOperators/Local.lean

```diff
@@ -0,0 +1,341 @@
+/-
+Copyright (c) 2025 Bryan Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, Bryan Wang
+-/
+import FLT.AutomorphicForm.QuaternionAlgebra.HeckeOperators.Abstract -- abstract Hecke ops
+import FLT.AutomorphicForm.QuaternionAlgebra.Defs -- definitions of automorphic forms
+import FLT.QuaternionAlgebra.NumberField -- rigidifications of quat algs
+import FLT.DedekindDomain.FiniteAdeleRing.LocalUnits -- for (π 0; 0 1)
+import Mathlib.Data.Matrix.Reflection
+
+open NumberField IsQuaternionAlgebra.NumberField IsDedekindDomain
+open TotallyDefiniteQuaternionAlgebra
+open IsDedekindDomain.HeightOneSpectrum
+open scoped TensorProduct
+open scoped Pointwise
+
+namespace TotallyDefiniteQuaternionAlgebra.WeightTwoAutomorphicForm.HeckeOperator
+
+-- let F be a (totally real) number field
+variable {F : Type*} [Field F] [NumberField F]
+
+namespace Local
+
+variable {v : HeightOneSpectrum (𝓞 F)}
+
+variable (α : v.adicCompletionIntegers F)
+
+variable (hα : α ≠ 0)
+
+-- Not sure if this could be somewhere else.
+variable (v) {α hα} in
+lemma valuation_lt_one_iff_mem_maximalIdeal {t : adicCompletionIntegers F v} :
+    Valued.v (t : adicCompletion F v) < 1
+    ↔ t ∈ IsLocalRing.maximalIdeal (adicCompletionIntegers F v) :=
+  Iff.trans
+  (Valuation.isEquiv_iff_val_lt_one.mp
+    (Valuation.isEquiv_valuation_valuationSubring Valued.v))
+  (ValuationSubring.valuation_lt_one_iff (adicCompletionIntegers F v) _).symm
+
+variable (v) {α hα} in
+/-- The subgroup `U1 v = GL2.localTameLevel v`. -/
+noncomputable abbrev U1 : Subgroup (GL (Fin 2) (adicCompletion F v)) := (GL2.localTameLevel v)
+
+/- Some lemmas in this section could be placed somewhere else in greater generality. -/
+namespace GL2
+
+/-- The matrix element `diag[α, 1]`. -/
+noncomputable abbrev diag : (GL (Fin 2) (adicCompletion F v)) :=
+  Matrix.GeneralLinearGroup.diagonal (![⟨(α : v.adicCompletion F),
+    (α : v.adicCompletion F)⁻¹, by
+      rw [mul_inv_cancel₀]
+      exact_mod_cast hα, by
+      rw [inv_mul_cancel₀]
+      exact_mod_cast hα⟩, 1])
+
+lemma diag_def :
+    (diag α hα : Matrix (Fin 2) (Fin 2) (adicCompletion F v))
+    = !![↑α, 0; 0, 1] := by
+  rw[diag, Matrix.GeneralLinearGroup.diagonal]
+  ext i j; fin_cases i; all_goals fin_cases j
+  all_goals simp
+
+lemma conjBy_diag {a b c d : adicCompletion F v} :
+    (diag α hα)⁻¹ * !![a, b; c, d] * diag α hα
+    = !![a, (α : v.adicCompletion F)⁻¹ * b; c * α, d] := by
+  simp only [Matrix.coe_units_inv, diag_def, Matrix.inv_def, Matrix.det_fin_two_of, mul_one,
```

Comment:
This is fine for FLT, but for mathlib my instinct would be to tell you to make API for `Matrix.GeneralLinearGroup.diagonal` e.g. `diagonal_mul_apply` and `mul_diagonal_apply` and get this proof to be something like
```
  ext i j
  fin_cases i; fin_cases j; simp; ring
```